statefun-docs/docs/concepts/index.rst

```diff
@@ -0,0 +1,97 @@
+
+.. contents:: :local:
+
+Event Ingress
+=============
+
+Stateful Function applications sit squarely in the event driven space, so the natural place to start is with getting events into the system.
+
+.. figure:: ../_static/images/concepts/statefun-app-ingress.svg
+    :width: 85%
+    :align: center
+
+In stateful functions, the component that ingests records into the system is called an event ingress.
+This can be anything from a Kafka topic, to a messsage queue, to an http request.
+Anything that can get data into the system and trigger the intitial functions to begin compution.
+
+Stateful Functions
+==================
+
+At the core of the diagram are the namesake stateful functions.
+
+.. figure:: ../_static/images/concepts/statefun-app-functions.svg
+    :width: 85%
+    :align: center
+
+Think of these as the building blocks for your service.
+They can message each other arbitrarily, which is one way in which this framework moves away from the traditional stream processing view of the world.
+Instead of building up a static dataflow DAG, these functions can communicate with each other in arbitrary, potentially cyclic, even round trip ways.
+
+If you are familiar with actor programming, this does share certain similarities in its ability to dynamically message between components.
+However, there are a number of significant differences.
+
+Persisted Values
+================
+
+The first is that all functions have locally embedded state, known as persisted values.
+
+.. figure:: ../_static/images/concepts/statefun-app-state.svg
+    :width: 85%
+    :align: center
+
+One of Apache Flink's core strengths is its ability to provide fault-tolerant local state.
+When inside a function, while it is performing some computation, you are always working with local state in local variables.
+
+Fault Tolerance
+===============
+
+For both state and messaging, Stateful Function's is still able to provide the exactly once guaruntees users expect from a modern data processessing framework.
+
+.. figure:: ../_static/images/concepts/statefun-app-fault-tolerance.svg
+    :width: 85%
+    :align: center
+
+In the case of failure, the entire state of the world (both persisted values and messages) are rolled back to simulate completely failure free execution.
+
+These guaruntees are provided with no database required, instead Stateful Function's leverages Apache Flink's proven snapshotting mechanism.
+
+Event Egress
+============
+
+Finally, applications can output data to external systems via event egress's.
+
+.. figure:: ../_static/images/concepts/statefun-app-egress.svg
+    :width: 85%
+    :align: center
+
+Of course, functions perform arbitrary computation and can do whatever they like, which includes making RPC calls and connecting to other systems.
+By using an event egress, applications can leverage pre-built interfaces and data output is tied into the systems fault tolerance to avoid data loss.
```

Comment:
Does this sentence imply that making arbitrary RPC calls / connecting to other systems can have data loss, without using event egresses? It sort of read to me like that, it being after the `Of course, functions perform arbitrary ...` sentence.

But that isn't true, as using async operations we still guarantee at-least once interactions with external systems.

Reply:
I'll reword